stdlib/event/event.lua

```diff
@@ -67,48 +76,71 @@ function Event.register(event, handler)
 end
 
 --- Calls the registerd handlers
--- @param event LuaEvent as created by script.raise_event
-function Event.dispatch(event)
-    Core.fail_if_missing(event, "missing event argument")
-    if Event._registry[event.name] then
-        for _, handler in pairs(Event._registry[event.name]) do
-            local metatbl = { __index = function(tbl, key) if key == '_handler' then return handler else return rawget(tbl, key) end end }
-            setmetatable(event, metatbl)
-            local success, err = pcall(handler, event)
-            if not success then
-                -- may be nil in on_load
-                if _G.game then
-                    if Game.print_all(err) == 0 then
-                        -- no players received the message, force a real error so someone notices
-                        error(err)
+-- Will stop dispatching remaning handlers if any handler passes invalid event data.
+-- Handlers are dispatched in the order they were created
+-- @tparam table event LuaEvent as created by script.raise_event
+-- @see https://forums.factorio.com/viewtopic.php?t=32039#p202158 Invalid Event Objects
+function Event.dispatch(event) --#70
+    if event then
+        local name = event.input_name or event.name  --#76
+        if Event._registry[name] then
+            local force_crc = Event.force_crc --#74
+            for idx, handler in ipairs(Event._registry[name]) do
+                -- Check for userdata and stop processing further handlers if not valid
+                for _, val in pairs(event) do --#78
+                    if type(val) == "table" and val.__self == "userdata" and not val.valid then
+                        return
                     end
-                else
-                    -- no way to handle errors cleanly when the game is not up
-                    error(err)
                 end
-                return
-            end
-            if err then
-                return
+
+                local metatbl = { __index = function(tbl, key) if key == '_handler' then return handler else return rawget(tbl, key) end end }
+                setmetatable(event, metatbl)
```

Comment:
This metatable is probably overkill -- if fields can simply be set directly on events, let's just set `_handler`. I wrote this when I thought setting fields was bad, but that doesn't seem like a valid concern anymore.